Source/levels/gendung.cpp

```diff
@@ -95,83 +95,46 @@ std::optional<Size> GetSizeForThemeRoom(int floor, Point origin, int minSize, in
 	if (origin.x + maxSize > DMAXX && origin.y + maxSize > DMAXY) {
 		return {}; // Original broken bounds check, avoids lower right corner
 	}
-	if (origin.x + minSize > DMAXX || origin.y + minSize > DMAXY) {
-		return {}; // Skip definit OOB cases
-	}
 	if (IsNearThemeRoom(origin)) {
 		return {};
 	}
 
-	bool yFlag = true;
-	bool xFlag = true;
-	int xCount = 0;
-	int yCount = 0;
-
-	int xArray[20] = {};
-	int yArray[20] = {};
-
-	for (int ii = 0; ii < maxSize; ii++) {
-		if (xFlag && origin.y + ii < DMAXY) {
-			for (int xx = origin.x; xx < origin.x + maxSize && xx < DMAXX; xx++) {
-				if (dungeon[xx][origin.y + ii] != floor) {
-					if (xx >= /* origin.x + */ minSize) {
-						break;
-					}
-					xFlag = false;
-				} else {
-					xCount++;
-				}
-			}
-			if (xFlag) {
-				xArray[ii] = xCount;
-				xCount = 0;
-			}
-		}
-		if (yFlag && origin.x + ii < DMAXX) {
-			for (int yy = origin.y; yy < origin.y + maxSize && yy < DMAXY; yy++) {
-				if (dungeon[origin.x + ii][yy] != floor) {
-					if (yy >= /* origin.y + */ minSize) {
-						break;
-					}
-					yFlag = false;
-				} else {
-					yCount++;
-				}
-			}
-			if (yFlag) {
-				yArray[ii] = yCount;
-				yCount = 0;
+	const int maxWidth = std::min(maxSize, DMAXX - origin.x);
+	const int maxHeight = std::min(maxSize, DMAXY - origin.y);
+
+	Size room { maxWidth, maxHeight };
+
+	for (int i = 0; i < maxSize; i++) {
+		int width = 0;
+		if (i < maxHeight) {
+			for (int x = 0; x < room.width; x++) {
+				if (dungeon[origin.x + x][origin.y + i] != floor)

```

Comment:
Arguable if it's any clearer but width and x are interchangable in this loop. Could just use width directly.
```suggestion
			for (; width < room.width;) {
				if (dungeon[origin.x + width][origin.y + i] != floor)
```

(same for the loop over y)

Reply:
Good catch, that is actually a nice improvement :)